Documentation/development-environment.md

```diff
@@ -0,0 +1,125 @@
+===============================================================================
+download : container_download | Download containers if pull is required or told to always pull (all nodes) - 192.44s
+kubernetes/preinstall : Update package management cache (YUM) --------- 178.26s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) - 102.24s
+docker : ensure docker packages are installed -------------------------- 57.20s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 52.33s
+kubernetes/preinstall : Install packages requirements ------------------ 25.18s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 23.74s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 18.90s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 15.39s
+kubernetes/master : Master | wait for the apiserver to be running ------ 12.44s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 11.83s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 11.66s
+kubernetes/node : install | Copy kubelet from hyperkube container ------ 11.44s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 11.41s
+download : container_download | Download containers if pull is required or told to always pull (all nodes) -- 11.00s
+docker : Docker | pause while Docker restarts -------------------------- 10.22s
+kubernetes/secrets : Check certs | check if a cert already exists on node --- 6.05s
+kubernetes-apps/network_plugin/flannel : Flannel | Wait for flannel subnet.env file presence --- 5.33s
+kubernetes/master : Master | wait for kube-scheduler -------------------- 5.30s
+kubernetes/master : Copy kubectl from hyperkube container --------------- 4.77s
+[leseb@tarox kubespray]$
+[leseb@tarox kubespray]$
+[leseb@tarox kubespray]$ vagrant ssh k8s-01
+Last login: Fri Jan 12 09:22:18 2018 from 192.168.121.1
+
+[vagrant@k8s-01 ~]$ kubectl get nodes
+NAME      STATUS    ROLES         AGE       VERSION
+k8s-01    Ready     master,node   2m        v1.9.0+coreos.0
+k8s-02    Ready     node          2m        v1.9.0+coreos.0
+k8s-03    Ready     node          2m        v1.9.0+coreos.0
+k8s-04    Ready     node          2m        v1.9.0+coreos.0
+k8s-05    Ready     node          2m        v1.9.0+coreos.0
+k8s-06    Ready     node          2m        v1.9.0+coreos.0
+k8s-07    Ready     node          2m        v1.9.0+coreos.0
+k8s-08    Ready     node          2m        v1.9.0+coreos.0
+k8s-09    Ready     node          2m        v1.9.0+coreos.0
+```
+
+## Development workflow
+
+Install git on k8s-01:
```

Comment:
what if we just have git and the rook repo on the host instead of inside the cluster? `kubectl` should be able to connect to the cluster from the host. And if you build rook on the host, you can push it to the docker registry inside the VM(s). It keeps the development environment independent from creating the VMs.

Reply:
I prefer this as well, because when you tear down your VMs and bring them up from scratch again, you wouldn't have to build again.  You can transfer a container image from your host with commands similar to the following:
```
docker save -o /tmp/tmp.docker rook/rook
for vm in ${vms} ; do vagrant scp /tmp/tmp.docker $vm:~/. ; done
for vm in ${vms} ; do vagrant ssh $vm -- docker load -i tmp.docker; done
for vm in ${vms} ; do vagrant ssh $vm -- docker tag rook/rook rook/rook:master; done
```
I've copied a full helper script to the following gist: https://gist.github.com/jbw976/3d61b87fdc8d81df57069a20b9e188b8

Reply:
Not sure why you'd ever scratch your cluster but yes that's probably more flexible this way. That's the path I had taken originally, I'll move again in that direction.